File: schemas/notifications-system/convex/notification_delivery_attempts.ts
// notification_delivery_attempts: Per-notification, per-channel delivery attempt log with full audit trail and retry tracking.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const notification_delivery_attempts = defineTable({
  notificationId: v.id("notifications"),
  channelId: v.id("notification_channels"),

  // Delivery lifecycle. Mutually exclusive, progresses forward.
  status: v.union(
    v.literal("pending"),
    v.literal("queued"),
    v.literal("sent"),
    v.literal("delivered"),
    v.literal("bounced"),
    v.literal("failed")
  ),

  providerMessageId: v.optional(v.string()), // e.g., SendGrid's "X-Message-Id", Twilio's "SM..." SID.
  providerResponse: v.optional(v.any()), // Raw provider response for debugging.

  errorCode: v.optional(v.string()), // Provider-specific error code.
  errorMessage: v.optional(v.string()), // Human-readable error description.

  attemptNumber: v.number(), // Which attempt this is (1 = first try, 2 = first retry, etc.).
  nextRetryAt: v.optional(v.number()), // When the next retry is scheduled. Null = no retry planned.

  sentAt: v.optional(v.number()), // When the provider accepted the request.
  deliveredAt: v.optional(v.number()), // When delivery was confirmed (from provider webhook).
  updatedAt: v.number(), // Tracks the latest status transition.
})
  .index("by_notification_id", ["notificationId"])
  .index("by_channel_id_status", ["channelId", "status"])
  .index("by_provider_message_id", ["providerMessageId"])
  .index("by_status_next_retry_at", ["status", "nextRetryAt"])
  .index("by_creation_time", ["_creationTime"]);

File: schemas/notifications-system/spacetimedb/notification_delivery_attempts.rs
// notification_delivery_attempts: Per-notification, per-channel delivery attempt log with full audit trail and retry tracking.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

// Delivery lifecycle status. Progresses forward; mutually exclusive.
// type: String
#[derive(SpacetimeType, Clone)]
pub enum DeliveryAttemptStatus {
    Pending,   // Attempt created, not yet processed.
    Queued,    // In the delivery queue, waiting to be sent.
    Sent,      // Handed off to the provider (provider accepted the request).
    Delivered, // Provider confirmed delivery to the recipient.
    Bounced,   // Provider reported a bounce (email bounced, push token invalid, etc.).
    Failed,    // Provider returned an error or max retries exhausted.
}

#[spacetimedb::table(name = notification_delivery_attempts, public)]
pub struct NotificationDeliveryAttempt {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub notification_id: String, // FK → notifications.id (cascade delete)

    #[index(btree)]
    pub channel_id: String, // FK → notification_channels.id (restrict delete)

    #[index(btree)]
    pub status: DeliveryAttemptStatus, // Default: Pending

    #[index(btree)]
    pub provider_message_id: Option<String>, // e.g., SendGrid "X-Message-Id", Twilio "SM..." SID.

    pub provider_response: Option<String>, // JSON — raw provider response for debugging.

    pub error_code: Option<String>,    // Provider-specific error code (e.g., "550", "InvalidRegistration").
    pub error_message: Option<String>, // Human-readable error description.

    pub attempt_number: i32, // Which attempt this is (1 = first try, 2 = first retry, etc.). Default: 1

    pub next_retry_at: Option<Timestamp>, // When the next retry is scheduled. Null = no retry planned.

    pub sent_at: Option<Timestamp>,      // When the provider accepted the request.
    pub delivered_at: Option<Timestamp>,  // When delivery was confirmed (from provider webhook).

    #[index(btree)]
    pub created_at: Timestamp,

    pub updated_at: Timestamp, // Tracks the latest status transition.
}
